src/Domain/SegmentId.hpp

```diff
@@ -0,0 +1,115 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines class SegmentId.
+
+#pragma once
+
+#include <cstdint>
+#include <iosfwd>
+#include <pup.h>
+
+#include "Domain/Side.hpp"
+#include "ErrorHandling/Assert.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+
+/// \ingroup ComputationalDomain
+/// A SegmentId labels a segment of the interval [-1,1].
+/// A SegmentId is used to (partially) identify an Element.
+/// The segment is the interval [-1 + 2*(N/D), -1 + 2*(N+1)/D]
+/// where N = index_ and D = 2 ^ refinement_level_.
+class SegmentId {
+ public:
+  /// Default constructor needed for Charm++ serialization.
+  SegmentId() = default;
+
+  SegmentId(size_t refinement_level, size_t index);
+
+  size_t refinement_level() const noexcept { return refinement_level_; }
+
+  size_t index() const noexcept { return index_; }
+
+  SegmentId id_of_parent() const;
+
+  SegmentId id_of_child(Side side) const;
+
+  /// The id this segment would have if the coordinate axis were flipped.
+  SegmentId id_if_flipped() const;
+
+  /// The logical coordinate of the endpoint of the segment on the given Side.
+  double endpoint(Side side) const;
+
+  /// The logical coordinate of the midpoint of the segment
+  double midpoint() const {
+    return -1.0 + (1.0 + 2.0 * index_) / two_to_the(refinement_level_);
+  }
+
+  /// Serialization for Charm++
+  void pup(PUP::er& p);  // NOLINT
```

Comment:
should check that `pup` works correctly. See `TestHelpers.hpp`, which has a `serialize_and_deserialize` function.